bigtable/client/rpc_metadata_holder.h

```diff
@@ -0,0 +1,220 @@
+ * The value of x-goog-request-params starts with one of the following suffix
+ *    "parent=" : Operation in instance, e.g. TableAdmin::CreateTable.
+ *    "table_name=" : table_id is known at the time of creation, e.g.
+ * Table::Apply.
+ *    "name=" : this is used when table|_id is known only in the RPC call, e.g.
+ * TableAdmin::GetTable.
+ *
+ */
+enum RPCRequestParamType { kParent, kName, kTableName };
+
+class GoogleCloudResourcePrefix : public RPCMetadataPair {
+ public:
+  explicit GoogleCloudResourcePrefix(std::string resource_string) {
+    set_key(std::move(kKey));
+    set_value(std::move(resource_string));
+  }
+
+ private:
+  std::string const kKey = "google-cloud-resource-prefix";
+};
+
+class XGoogleRequestParamas : public RPCMetadataPair {
+ public:
+  explicit XGoogleRequestParamas(std::string resource_string,
+                                 RPCRequestParamType request_param_type) {
+    std::string value;
+    switch (request_param_type) {
+      case RPCRequestParamType::kParent:
+        value = kValuePrefixParent;
+        value += resource_string;
+        break;
+      case RPCRequestParamType::kName:
+        value = kValuePrefixName;
+        value += resource_string;
+        break;
+      case RPCRequestParamType::kTableName:
+        value = kValuePrefixTableName;
+        value += resource_string;
+        break;
+    }
+
+    set_key(std::move(kKey));
+    set_value(std::move(value));
+  }
+
+  explicit XGoogleRequestParamas(std::string resource_string,
+                                 RPCRequestParamType request_param_type,
+                                 std::string table_id) {
+    std::string value;
+    switch (request_param_type) {
+      case RPCRequestParamType::kParent:
+        value = kValuePrefixParent;
+        value += resource_string;
+        value += "/tables/" + table_id;
+        break;
+      case RPCRequestParamType::kName:
+        value = kValuePrefixName;
+        value += resource_string;
+        value += "/tables/" + table_id;
+        break;
+      case RPCRequestParamType::kTableName:
+        value = kValuePrefixTableName;
+        value += resource_string;
+        value += "/tables/" + table_id;
+        break;
+    }
+    set_key(std::move(kKey));
+    set_value(std::move(value));
+  }
+
+ private:
+  std::string const kKey = "x-goog-request-params";
+  static constexpr const char* kValuePrefixParent = "parent=";
+  static constexpr const char* kValuePrefixName = "name=";
+  static constexpr const char* kValuePrefixTableName = "table_name=";
+};
+
+/// RPCMetadataHolder holds supported metadata
+class RPCMetadataHolder {
+ public:
+  /**
+   * Constructor with default metadata pair.
+   *
+   * @param resource_name hierarchical name of resource, including  project id,
+   * instance id
+   *        and/or table_id.
+   * @param request_param_type type to decide prefix for the value of
+   * x-goog-request-params
+   */
+  RPCMetadataHolder(std::string resource_name,
```

Comment:
The other policies have an abstract base class, and one or more concrete implementations.  The expectation is that we or our users will create new classes that implement the `RPCBackoffPolicy` interface (for example).

This is only a concrete implementation.  If we only have a concrete implementation then we do not need `unique_ptr`, and `clone` and all the other machinery to preserve state and behavior.

So, which one should we implement?  I cannot think of an scenario where a user or us need to customize the behavior of this class.  Therefore we should treat it as a concrete class.  That means we do not need `clone()`, we do not need `cloneWithModifications()`, we do not need `DefaultRPCMetadataHolder()`.

Just create an object of this class as a member variable, pass by (const) reference when needed, use copy constructors if needed too.

Reply:
Yes, a concrete implementation would be better, I will remove unique_ptr and clone.
But we do need MakeNew as x-goog-request-params changes its value wherever table_id is part of request. So we do need 2 behaviours, eager or default and lazy whenever table_id is part of request.

Reply:
> But we do need MakeNew as x-goog-request-params changes its value wherever table_id
> is part of request. So we do need 2 behaviours, eager or default and lazy whenever table_id
> is part of request.

That is a good point.  Maybe the `bigtable::TableAdmin` class can create an instance on the stack for the cases where a `table_id` is needed?

Reply:
changed